src/internal.c

```diff
@@ -6237,6 +6237,9 @@ int SetSSL_CTX(WOLFSSL* ssl, WOLFSSL_CTX* ctx, int writeDup)
     ssl->ctx     = ctx; /* only for passing to calls, options could change */
     /* Don't change version on a SSL object that has already started a
      * handshake */
+#if defined(WOLFSSL_HAPROXY)
+    ssl->initial_ctx     = ctx; /* Save access to session key materials */
+#endif
```

Comment:
The reference counter needs to be incremented twice on `ctx`. Be sure to call  `wolfSSL_CTX_free` on `ssl->initial_ctx` too.